polaris-service/src/main/java/org/apache/polaris/service/catalog/io/WasbTranslatingFileIO.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.service.catalog.io;
+
+import java.util.Map;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.polaris.core.storage.StorageLocation;
+import org.apache.polaris.core.storage.azure.AzureLocation;
+
+/**
+ * A {@link FileIO} implementation that translates WASB paths into ABFS paths and then delegates to
+ * another underlying FileIO implementation
+ */
+public class WasbTranslatingFileIO implements FileIO {
+  private final FileIO io;
+
+  private static final String WASB_SCHEME = "wasb";
+  private static final String ABFS_SCHEME = "abfs";
+
+  public WasbTranslatingFileIO(FileIO io) {
```

Comment:
Since it looks like the current structure of FileIOFactory is forcing this to wrap all FileIOs including non-Azure-related ones, we might as well make the class itself more generalized and we could move the Azure-specificity into config.

What if we just called this thing `SchemeTranslatingFileIO` that takes a map of source schemes to destination schemes? And get the map through config

Reply:
Oh I see you also have the endpoint rewrite from `blob` to `dfs` which wouldn't fit into that. I agree we might not want to go down the road of overly general regexes either.

Per https://github.com/apache/iceberg/issues/10127#issuecomment-2080302481 the underlying Azure SDK doesn't actually seem to actually care about the endpoint and will internally sort into both a `dfs` and a `blob` client, so in theory a scheme-only replacement should still work, but I guess it could be more fragile.

@collado-mike What do you think? Should we just keep this very wasb-specific?

Reply:
> it looks like the current structure of FileIOFactory is forcing this to wrap all FileIOs including non-Azure-related ones

It's doing a little more than just scheme translating, and this is only true if Polaris is actually configured by an admin to use this factory.

Reply:
If I were designing this, I would have a `TransformingFileIO` with a constructor like 
```java
public TransformingFileIO(Function<String, String> pathTransformer, FileIOFactory delegate) {
}
```
and the azure class would extend that class and pass in a function that did the wasb->abfs conversion. But that's the kind of code refactoring that can be done later without any compatibility or behavior impact, so... 🤷🏽‍♂️